core/src/test/scala/unit/kafka/coordinator/transaction/TransactionStateManagerConcurrencyTest.scala

```diff
@@ -0,0 +1,158 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package kafka.coordinator.transaction
+
+import kafka.coordinator.AbstractCoordinatorConcurrencyTest
+import kafka.coordinator.transaction.TransactionStateManagerConcurrencyTest._
+import kafka.coordinator.AbstractCoordinatorConcurrencyTest._
+import kafka.utils.timer.MockTimer
+import kafka.utils.{ MockScheduler, Pool, TestUtils}
+
+import org.apache.kafka.common.TopicPartition
+import org.apache.kafka.common.internals.Topic.TRANSACTION_STATE_TOPIC_NAME
+import org.apache.kafka.common.protocol.Errors
+import org.apache.kafka.common.requests.TransactionResult
+import org.apache.kafka.common.utils.MockTime
+
+import org.easymock.EasyMock
+import org.junit.Assert.{ assertEquals, assertFalse, assertTrue }
+import org.junit.{ After, Before, Test }
+
+import scala.collection.Map
+import scala.collection.mutable
+import scala.collection.JavaConverters._
+
+class TransactionStateManagerConcurrencyTest extends AbstractCoordinatorConcurrencyTest[Transaction] {
```

Comment:
Out of curiosity, why do this at the lower level instead of the way you did the group coordinator tests?

Reply:
@hachikuji  Thank you for the review. I was trying to use logic from existing mock tests in both cases to trigger the operations. The transaction coordinator unit tests use a mock `TransactionStateManager` and various other mocks, so it looked like a lot more work to get those to trigger the path with `appendRecords`. But it does make sense to write the tests at the coordinator level, so I will update.